### PR TITLE
sink(ticdc): Implement prep stmt cache in mysql sink

### LINKED ISSUE #7342: CDC should use prepared statement rather than query command to submit sql to downstream TiDB
### Is your feature request related to a problem?

using query command can not hit TiDB plan cache, causing high TiDB CPU and high parse/compile time, from a performance perspective, I would like to propose using the prepared statement.

image

### Describe the feature you'd like

please refer to Performance Tuning Practices for OLTP Scenarios
https://docs.pingcap.com/tidb/stable/performance-tuning-practices

### Describe alternatives you've considered

no alternatives currently.

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by dbsid**
### What problem does this PR solve?

Issue Number: close #7342

### What is changed and how it works?
The approach to achieve this goal is to modify the CDC code to use prepared statements instead of query commands when submitting SQL to TiDB. The implementation involves creating a LRU cache for prepared statements which are frequently used SQL queries, and reusing them for subsequent operations.
To implement this, I follow the steps below:
1. Identify frequently used SQL queries in CDC communication with TiDB.
2. Use the driver's DB.Prepare function to prepare and cache these statements.
3. Use the cached prepared statements in subsequent operations instead of sending raw query commands.
4. Use the Tx.Stmt function to copy the prepared statements to the transaction scope if the statements are used in transactions.
5. A new sink option prep-prep-stmt-cache-size is added to control the size of LRU cache.

#### Tests 

 - [x]Unit test
 - [x]Integration test
 - [x]Manual test (add detailed scripts or steps below)
 - No code

##### Will it cause performance regression or break compatibility?
performance improvement is expected.
With sysbench 50 threads, the stmt cache new feature:
  1. CDC checkpoint lag is reduced from 52.9s to 0.52s, cdc memory usage is reduced from 4.9GB to 0.89GB.
  2. CDC average transaction sink flush duration is reduced from 115ms to 68ms.
  3. For downstream TiDB cluster, plan cache hits rate increase from 0 to 100%, avg the query duration is reduced from 740us to 640us, TiDB instance cpu usage is reduced from 1147% to 960%.
  
image
image
image

##### Do you need to update user documentation, design documentation or monitoring documentation?
Yes
### Release note 

```release-note
Two new sink URL options cache-prep-stmts and prep-stmt-cache-size are added to enable the new feature prep stmt cache and control the size of LRU cache.
```

### CODE REVIEW

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 1]
```diff
@@ -99,8 +102,15 @@ func NewMySQLBackends(
 		return nil, err
 	}
 
-	db.SetMaxIdleConns(cfg.WorkerCount)
-	db.SetMaxOpenConns(cfg.WorkerCount)
+	db.SetMaxIdleConns(cfg.WorkerCount + 1)
+	db.SetMaxOpenConns(cfg.WorkerCount + 1)
```
Comment (amyangfei) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:106]: Please add a comment about why plus one here

File: pkg/sink/mysql/config.go [hunk 2]
```diff
@@ -332,3 +342,26 @@ func getBatchDMLEnable(values url.Values, batchDMLEnable *bool) error {
...
+		return cerror.WrapError(cerror.ErrMySQLInvalidConfig, err)
+	}
+	if c <= 0 {
+		return cerror.WrapError(cerror.ErrMySQLInvalidConfig,
+			fmt.Errorf("invalid prep-stmt-cache-size %d, which must be greater than 0", c))
+	}
+	if c > maxPrepStmtCacheSize {
+		log.Warn("prep-stmt-cache-size too large",
+			zap.Int("original", c), zap.Int("override", maxPrepStmtCacheSize))
+		c = maxMaxTxnRow
```
Comment (amyangfei) [inline pkg/sink/mysql/config.go:363]: Should this be `maxPrepStmtCacheSize`

File: pkg/sink/mysql/config_test.go [hunk 3]
```diff
@@ -245,6 +246,11 @@ func TestParseSinkURIOverride(t *testing.T) {
 		checker: func(sp *Config) {
 			require.EqualValues(t, sp.tidbTxnMode, defaultTiDBTxnMode)
 		},
+	}, {
+		uri: "mysql://127.0.0.1:3306/?prep-stmt-cache-size=1000000", // int32 max
```
Comment (amyangfei) [inline pkg/sink/mysql/config_test.go:250]: Does it mean to test a size that is larger than `maxPrepStmtCacheSize`?

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ import (
 	"time"
 
 	dmysql "github.com/go-sql-driver/mysql"
+	lru "github.com/hashicorp/golang-lru"
```
Comment (amyangfei) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:25]: Note the license of `hashicorp/golang-lru` is MPL, we need to confirm whether it is appropriate to use in tiflow repo.

**Comment by amyangfei** [comment]
Besides some unit tests in `cdc/sinkv2/eventsink/txn/mysql` module need to be fixed, since more prepare statements will be sent in database connection.

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 1 above)
Comment (dbsid) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:106]: got it, will followup

File: pkg/sink/mysql/config.go (hunk 2 above)
Comment (dbsid) [reply pkg/sink/mysql/config.go:363]: yes, will correct it.

File: pkg/sink/mysql/config_test.go (hunk 3 above)
Comment (dbsid) [reply pkg/sink/mysql/config_test.go:250]: yes, will correct it

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 4 above)
Comment (dbsid) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:25]: agree, seems need to seek for advice from the legal colleague

**Comment by sdojjy** [comment]
the release note is wrong , should be `prep-stmt-cache-size`

File: cdc/sinkv2/eventsink/txn/mysql/mysql_test.go [hunk 5]
```diff
@@ -288,7 +288,8 @@ func TestNewMySQLBackendExecDML(t *testing.T) {
 	defer cancel()
 	changefeed := "test-changefeed"
 	contextutil.PutChangefeedIDInCtx(ctx, model.DefaultChangeFeedID(changefeed))
-	sinkURI, err := url.Parse("mysql://127.0.0.1:4000/?time-zone=UTC&worker-count=1")
+	sinkURI, err := url.Parse(
+		"mysql://127.0.0.1:4000/?time-zone=UTC&worker-count=1&cache-prep-stmts=false")
```
Comment (amyangfei) [inline cdc/sinkv2/eventsink/txn/mysql/mysql_test.go:292]: Better to leave a TODO to test txn sink behavior when `cache-prep-stmts` is true

Comment (dbsid) [reply cdc/sinkv2/eventsink/txn/mysql/mysql_test.go:296]: the TODO is added

**Review by amyangfei — APPROVED** [review]

**Review by hi-rustin — APPROVED** [review]

**Comment by hi-rustin** [comment]
Some tests failed.

**Review by hi-rustin — APPROVED** [review]

**Pull request merged**